### PR TITLE
Fix post featured image scaling on the front

### LINKED ISSUE #42322: Featured image: height set is not being inherited correctly on the frontend
### Description

This is a tricky one to find. A featured image block with a set height will show correctly in the editor but the frontend will display something different. There's a couple of reasons to this: 
- the img tag is not getting the height applied to the featured image block wrapper
- the editor has a different markup than the frontend (the anchor link when the featured image is a link is missing in the editor but present in the frontend)

### Step-by-step reproduction instructions

On empty theme, paste this code on a page:

```
<!-- wp:query {"queryId":33,"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:columns {"backgroundColor":"cyan-bluish-gray"} -->
<div class="wp-block-columns has-cyan-bluish-gray-background-color has-background"><!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"><!-- wp:post-title {"isLink":true,"fontSize":"x-large"} /-->

<!-- wp:post-excerpt {"fontSize":"large"} /--></div>
<!-- /wp:column -->

<!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"><!-- wp:post-featured-image {"isLink":true,"height":"100%","align":"wide"} /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

Ensure that there's an image on the post you are displaying and the text forces the image to stretch.

Check the frontend, the image is not stretching there.

### Screenshots, screen recording, code snippet

Editor:

Screenshot 2022-07-11 at 12 32 46

Frontend:

Screenshot 2022-07-11 at 12 32 35

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by carolinan**
## What?

In the block editor, the post featured image `<figure>` wrapper has a set height. This height was missing on the front, so the image did not look the same in the editor and front.
This pr adds the height to the front, both to the `<figure>` and the `<a>` which is used when the image is linked.

Closes [#42322](https://github.com/WordPress/gutenberg/issues/42322)
## Why?

The image did not look the same in the editor and front when image height and scaling was set in the block settings.

## How?

Updates index.php and adds the height to the `get_block_wrapper_attributes` and the `<a>` when applicable.

## Testing Instructions
When testing this we need to take extra care to make sure that the border and border radius settings still work.

1. Make sure that your WordPress install has a few recent posts with a lot of text content and a featured image. - This is needed because we are going to use a query loop and use the long content to stretch the image.
2. Create a new post.
3. Copy the example code from the issue into the post:
```
<!-- wp:query {"queryId":33,"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:columns {"backgroundColor":"cyan-bluish-gray"} -->
<div class="wp-block-columns has-cyan-bluish-gray-background-color has-background"><!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"><!-- wp:post-title {"isLink":true,"fontSize":"x-large"} /-->

<!-- wp:post-excerpt {"fontSize":"large"} /--></div>
<!-- /wp:column -->

<!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"><!-- wp:post-featured-image {"isLink":true,"height":"100%","align":"wide"} /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```
3. The featured image should have a 100% height and "scale" set to cover, and stretch the height of the content.
4. Save and view the front. Confirm that the image looks the same on the front.

5. Continue to test the remaining settings. Confirm that the image looks correct on the front for each step.
For example:
- Enable and disable the linked image setting
- Set the image width to 45px
- Change the scale to contain and fill (The scale settings only show if width or height are adjusted).
- Change the border styles including radius
- Add an overlay

### CODE REVIEW

File: packages/block-library/src/post-featured-image/index.php [hunk 1]
```diff
@@ -49,22 +49,27 @@ function render_block_core_post_featured_image( $attributes, $content, $block )
 	if ( $is_link ) {
 		$link_target    = $attributes['linkTarget'];
 		$rel            = ! empty( $attributes['rel'] ) ? 'rel="' . esc_attr( $attributes['rel'] ) . '"' : '';
+		$height         = ! empty( $attributes['height'] ) ? 'style="height:' . esc_attr( $attributes['height'] ) . '"' : '';
```
Comment (jorgefilipecosta) [inline packages/block-library/src/post-featured-image/index.php]: Given that we are escaping a style from the security point of view, we also need to call the function `safecss_filter_attr`. So here we should do esc_attr( safecss_filter_attr( 'height:' . esc_attr( $attributes['height'] ) . '"' ) )`

File: packages/block-library/src/post-featured-image/index.php [hunk 2]
```diff
@@ -49,22 +49,27 @@ function render_block_core_post_featured_image( $attributes, $content, $block )
...
-	$wrapper_attributes = empty( $attributes['width'] )
-		? get_block_wrapper_attributes()
-		: get_block_wrapper_attributes( array( 'style' => "width:{$attributes['width']};" ) );
-
+	$width  = ! empty( $attributes['width'] ) ? 'width:' . $attributes['width'] . ';' : '';
+	$height = ! empty( $attributes['height'] ) ? 'height:' . $attributes['height'] . ';' : '';
+	if ( ! $height && ! $width ) {
+		$wrapper_attributes = get_block_wrapper_attributes();
+	} else {
+		$wrapper_attributes = get_block_wrapper_attributes( array( 'style' => $width . $height ) );
```
Comment (jorgefilipecosta) [inline packages/block-library/src/post-featured-image/index.php:71]: We need an identical update to the one referred above.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @carolinan, thank you for fixing this issue 👍 
I left some comments that I think we need to address before the merge but other than that, the PR looks good.

File: packages/block-library/src/post-featured-image/index.php (hunk 1 above)
Comment (carolinan) [reply packages/block-library/src/post-featured-image/index.php]: Updated, thank you.

File: packages/block-library/src/post-featured-image/index.php [hunk 3]
```diff
@@ -49,22 +49,27 @@ function render_block_core_post_featured_image( $attributes, $content, $block )
 	if ( $is_link ) {
 		$link_target    = $attributes['linkTarget'];
 		$rel            = ! empty( $attributes['rel'] ) ? 'rel="' . esc_attr( $attributes['rel'] ) . '"' : '';
+		$height         = ! empty( $attributes['height'] ) ? 'style="' . esc_attr( safecss_filter_attr( 'height:' . esc_attr( $attributes['height'] ) ) ) . '"' : '';
```
Comment (aaronrobertshaw) [inline packages/block-library/src/post-featured-image/index.php]:
```suggestion
		$height         = ! empty( $attributes['height'] ) ? 'style="' . esc_attr( safecss_filter_attr( 'height:' . $attributes['height'] ) ) . '"' : '';
```

I don't think we need the inner `esc_attr` calls here.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for fixing this issue @carolinan!

It tests well for me when:
- setting various combinations of explicit or relative widths
- adding borders, including a radius
- adding an overlay

After adding the `safecss_filter_attr` calls, it looks like we have a couple of redundant `esc_attr` calls now that we can clean up. After that, this should be good to go, nice work ✨ 

| Editor | Frontend |
|---|---|
| Screenshot 2023-01-09 at 7 58 17 pm | Screenshot 2023-01-09 at 7 58 09 pm |

File: packages/block-library/src/post-featured-image/index.php (hunk 3 above)
Comment (carolinan) [reply packages/block-library/src/post-featured-image/index.php]: I copied this from the suggestions from @jorgefilipecosta.  Looking at the other places in the plugin where 
`safecss_filter_attr` is used, I am seeing both escaped and unescaped attributes values, so which is correct?

Comment (aaronrobertshaw) [reply packages/block-library/src/post-featured-image/index.php]: > I copied this from the suggestions

Understood. I think there was a small copy and paste error within that suggestion.

There are a few inconsistencies within the Gutenberg codebase around when we escape using `esc_attr` combined with `safecss_filter_attr`. Whether you escape the attribute before `safecss_filter_attr` or after, you'd only need one.

I'd recommend using `esc_attr( safecss_filter_attr( ... ) )` as it follows the principle of escaping as late as possible. This was recommended to me in feedback by contributors more well versed in security best practices than I.

Comment (carolinan) [reply packages/block-library/src/post-featured-image/index.php]: Then let's go with that :)

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the tweaks @carolinan 👍 

This is still testing well for me, and the code looks good. Nice work!

**Pull request merged**